src/main/java/de/rub/nds/modifiablevariable/Logger/CustomLayout.java

```diff
@@ -0,0 +1,562 @@
+            this.header = header;
+            return this;
+        }
+
+        public CustomLayout.Builder withFooter(String footer) {
+            this.footer = footer;
+            return this;
+        }
+
+        public CustomLayout build() {
+            if (this.configuration == null) {
+                this.configuration = new DefaultConfiguration();
+            }
+
+            return new CustomLayout(
+                    this.configuration,
+                    this.regexReplacement,
+                    this.pattern,
+                    this.patternSelector,
+                    this.charset,
+                    this.alwaysWriteExceptions,
+                    this.disableAnsi,
+                    this.noConsoleNoAnsi,
+                    this.header,
+                    this.footer);
+        }
+    }
+
+    private static class PatternSelectorSerializer
+            implements AbstractStringLayout.Serializer,
+                    AbstractStringLayout.Serializer2,
+                    LocationAware {
+        private final PatternSelector patternSelector;
+        private final RegexReplacement replace;
+
+        private PatternSelectorSerializer(
+                PatternSelector patternSelector, RegexReplacement replace) {
+            this.patternSelector = patternSelector;
+            this.replace = replace;
+        }
+
+        public String toSerializable(LogEvent event) {
+            StringBuilder sb = AbstractStringLayout.getStringBuilder();
+
+            String var3;
+            try {
+                var3 = this.toSerializable(event, sb).toString();
+            } finally {
+                AbstractStringLayout.trimToMaxSize(sb);
+            }
+
+            return var3;
+        }
+
+        public StringBuilder toSerializable(LogEvent event, StringBuilder buffer) {
+            PatternFormatter[] formatters = this.patternSelector.getFormatters(event);
+            int len = formatters.length;
+
+            for (int i = 0; i < len; ++i) {
+                formatters[i].format(event, buffer);
+            }
+
+            if (this.replace != null) {
+                String str = buffer.toString();
+                str = this.replace.format(str);
+                buffer.setLength(0);
+                buffer.append(str);
+            }
+
+            return buffer;
+        }
+
+        public boolean requiresLocation() {
+            return this.patternSelector instanceof LocationAware
+                    && ((LocationAware) this.patternSelector).requiresLocation();
+        }
+
+        public String toString() {
+            StringBuilder builder = new StringBuilder();
+            builder.append(super.toString());
+            builder.append("[patternSelector=");
+            builder.append(this.patternSelector);
+            builder.append(", replace=");
+            builder.append(this.replace);
+            builder.append("]");
+            return builder.toString();
```

Comment:
Using a string builder here is no longer state of the art. String Builders should be used when iterating and calling concatenations within a loop. In this case, a simple String.format() or concantenation should do the trick. The compiler will optimize it on its own.

Reply:
If this is the case shouldn't this class inherit from `PatternLayout` to avoid duplicate code? Or is the class marked as final?

Reply:
Yes the class is marked as final so we need the duplicated code to guarantee the same functionality.

Reply:
Should I use the StringBuilder or the String concatination?

Reply:
Okay, I understand the necessity to duplicate the code from Log4J. Anyway, we should adhere to coding conventions here and use String concatination and Override annotations.